### PR TITLE
Better error handling, centralize dyanomdb conns and tables

### PR DESCRIPTION
**Pull request by loicgasser**
With this PR everyone should have a .boto file to serve the application in dev mode.
Connections are created once and reused over time (best practice).
Only boto layer2 interface is used now.

### CODE REVIEW

**Comment by gjn** [comment]
Great simplification. I'll review in detail tomorrow.

This needs a rebase because of https://github.com/geoadmin/mf-chsdi3/pull/1472 that was just merged. We'll test this PR in detail after tomorrows deploy.

**Comment by loicgasser** [comment]
ok, rebased

**Comment by loicgasser** [comment]
boto updated and parse_ts function replaced as discussed.

**Comment by cedricmoullet** [comment]
@gjn @loicgasser @procrastinatio can we deploy this tomorrow ?

**Comment by loicgasser** [comment]
Apparently for next one

**Comment by loicgasser** [comment]
@procrastinatio could you have a look and merge if ok with you?

**Comment by procrastinatio** [comment]
I have already pretty much tested, but until it is deployed, we won't know for sure.

**Comment by gjn** [comment]
How do we load test this a little bit?

We're moving from a per-request connection object to a global connection object...

**Comment by loicgasser** [comment]
boto is wrapper around Amzon REST API, so it s fine.
http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html

**Comment by gjn** [comment]
i was more thinking if the boto connection object could handle concurrency (same object used in multiple threads in parallel). as this might happen under load. usually, you need to design your classes/objects specifically for. i did not look in documentation nor in source code.

File: chsdi/models/clientdata_dynamodb.py [hunk 1]
```diff
@@ -42,11 +44,39 @@
...
+    # TODO use profile instead when correctly installed
+    try:
+        conn = connect_s3(profile_name=profile_name)
+    except Exception as e:
+        raise exc.HTTPBadRequest('S3: Error during connection %s' % e)
+    return conn
+
+
+conn_dynamodb = _get_dynamodb_conn()
+conn_s3 = _get_s3_conn()
```
Comment (gjn) [inline chsdi/models/clientdata_dynamodb.py]: this is called at webapp start, correctly? if yes, raising httpbadrequest in the functuons does not really make sense. also, will it stip startup of application?

Comment (gjn) [inline chsdi/models/clientdata_dynamodb.py]: 'functions' and 'stop', sorry.

File: chsdi/views/files.py [hunk 2]
```diff
@@ -165,8 +107,11 @@ def __init__(self, request):
                 self.file_id = id
             try:
                 key = self.bucket.get_key(self.file_id)
-            except:
-                raise exc.HTTPInternalServerError('Cannot access file with id=%s' % self.file_id)
+            except S3ResponseError as e:
+                raise exc.HTTPInternalServerError('Cannot access file with id=%s: %s' % (self.file_id, e))
+            except Exception as e:
+                raise exc.HTTPInternalServerError('Cannot access file with id=%s: %s' % (self.file_id, e))
```
Comment (gjn) [inline chsdi/views/files.py]: both raised exceptions are exactly the same, so why having both?

**Comment by gjn** [comment]
according to http://ls.pwd.io/2013/06/parallel-s3-uploads-using-boto-and-threads-in-python/, boto connections are not thread safe (and using new connection for every request is very fast). the official doc doesn't specifically mention thread safety for connection. Also, on SO, they say that boto connections are not thread safe: http://stackoverflow.com/questions/19523680/is-boto-library-thread-safe

Unless I get other information, I consider this PR critical and it shouldn't have been merged.

**Comment by procrastinatio** [comment]
Hem, is there any difference between a dynamodb and a s3 connection ? we only did use the same approach for s3 as for dynamodb (which obviously work).

**Comment by gjn** [comment]
we have introduced 2 global objects that are not thread-safe that could be used in 2 different threads on 2 concurrent requests. the behaviour then is undefined. in the lucky case, the request fails, in the unlucky case, the application crashes. in any case, it's a bug.

**Comment by gjn** [comment]
workers in tileworkers are processes, right?

**Comment by loicgasser** [comment]
connect has to be called once. then we create new buckets coming from the same connection. it works like sqlalchemy... http://stackoverflow.com/questions/11622548/boto-connect-xxx-method-and-connection-pools

**Comment by loicgasser** [comment]
I can find other links that will tell you the same.

**Comment by loicgasser** [comment]
If we want to be on the safe side, I can of course change that and create a new conn for each request as it was before. no problem and takes no time.

**Comment by gjn** [comment]
read the first answer from your link. it says exactly what I'm saying.

an alternative is to use connectionpool class, where the doc explicitely assures thread safety.

currently, it's a bug and we should correct it.

**Comment by loicgasser** [comment]
connection object uses a pool such that each connection is used only by a single thread at a time.

But let s not be dramatic and change that for a good sleep.

**Comment by gjn** [comment]
could you tell me where you've read that, because i didn't find anything.

**Comment by loicgasser** [comment]
same link last answer. To be sure I should dig in the code and see if connect_s3 inherits/uses ConnectionPool class, it did nt use to be threadsafe so you re probably right to worry. Anyway will revert this part.

**Comment by loicgasser** [comment]
so connect_s3 is wrapping S3Connection which extends AWSConnection which uses ConnectionPool class. so not a bug and we re safe.

**Comment by gjn** [comment]
ouf, great. thanks alot for verifying and sorry for steering the pot.

since which version is it thread-safe!

**Comment by gjn** [comment]
we can discuss the other in-code comments on tuesday.

File: chsdi/views/files.py (hunk 2 above)
Comment (loicgasser) [inline chsdi/views/files.py]: I thought to distinguish errors coming from S3 and otther erros.
S3ResponseError (meaning a request has been sent)
and
Exception (any other error)
text could be different.

**Comment by loicgasser** [comment]
you were right to make me check it out. I just trusted what I red but for this kind of things it's better to double check :D

**Pull request merged**